src/mame/apple/apple2e.cpp

```diff
@@ -4773,7 +4919,7 @@ static INPUT_PORTS_START( apple2efr )   // French AZERTY keyboard (Apple uses th
 	PORT_BIT(0x008, IP_ACTIVE_HIGH, IPT_UNUSED)
 	PORT_BIT(0x010, IP_ACTIVE_HIGH, IPT_UNUSED)
 	PORT_BIT(0x020, IP_ACTIVE_HIGH, IPT_UNUSED)
-	PORT_BIT(0x040, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_BACKSLASH)  PORT_CHAR(0xa3) PORT_CHAR('`')  // UK pound (actually to the left of the return key on the QSDF row)
+	PORT_BIT(0x040, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_TILDE)      PORT_CHAR('>') PORT_CHAR('<')   // actually the key between left shift and W
```

Comment:
You’re supposed to use `KEYCODE_BACKSLASH2` for this key.

Reply:
My last commit adds KEYCODE_BACKSLASH2, but without removing KEYCODE_BACKSLASH for all ISO layouts (UK, FR, and ES).  The reason I did not remove KEYCODE_BACKSLASH is that KEYCODE_BACKSLASH2 does not exist as a key on ANSI keyboards, and in the case of Apple IIe/IIc machines, the ISO keyboard has the same number of keys as the North American keyboard - unlike PCs where the ISO keyboard has one extra key.  On the Apple IIe/IIc, it is just two keys get swapped in physical position and in how they get encoded in the ROM.  Please advise if having both backslashes mapped is correct.

Regarding taking this PR out of draft mode, the main thing I have not done yet is create a new keyboard layout for the UK variant of the Platinum Apple IIe.  For that, I wanted to discuss a long-term design for the International Apple computers.
- My original thought was that they could all share the same PORT_CODE/PORT_CHAR defined layout as the US variant.  However,  `~ and \| encodings are swapped between North American and ISO layouts make that design impossible.
- A similar design to the above would be to have 2 US layouts (ISO US layout for UK, France, Italy, Germany, Sweden, Spain and North American US layout for US, Canada, and Latin American machines)
- Would MAME users expect to see the ''Input Assignments (this system)" to show the localized keyboard, or is having the US layout shown an option - given that the real Apple IIe/IIc machines have selection switches to pick either of the two layouts?  This would impact which of the two Apple IIe/c keyboard layouts will work when the keyboard mode is set to 'natural'.   Currently, the localized one works in 'natural' mode, but the US one will work instead if the above change is made.

Reply:
No, `KEYCODE_BACKSLASH` is present on ISO keyboards.

Using this image for reference:
![image](https://github.com/user-attachments/assets/409f7a2a-8f55-4c82-8d74-06295becb964)

* `KEYCODE_BACKSLASH` is for the key immediately to the left of return on the home row, with the £ pound sign on it.
* `KEYCODE_BACKSLASH2` is for the key immediately to the right of the left shift key on the bottom row, with angle brackets on it.

The key corresponding to `KEYCODE_BACKSLASH2` on ISO keyboards simply doesn’t exist on ANSI or JIS keyboards.  Users need to reassign it manually.  Assigning `KEYCODE_BACKSLASH2` and `KEYCODE_BACKSLASH` to that key by default will not do what people expect.

Reply:
> My original thought was that they could all share the same PORT_CODE/PORT_CHAR defined layout as the US variant. However, `~ and | encodings are swapped between North American and ISO layouts make that design impossible.

And that’s a bad idea, and not how we do things in MAME.  But remember `PORT_MODIFY` is a thing so you can reduce duplication if the matrix is mostly the same.

> Would MAME users expect to see the ''Input Assignments (this system)" to show the localized keyboard, or is having the US layout shown an option - given that the real Apple IIe/IIc machines have selection switches to pick either of the two layouts? This would impact which of the two Apple IIe/c keyboard layouts will work when the keyboard mode is set to 'natural'. Currently, the localized one works in 'natural' mode, but the US one will work instead if the above change is made.

No, always showing the US layout is not what users would expect.  For systems with plug-in keyboards, we provide options for language variants known to exist, even if they only differ in key cap labels, e.g. see these files:
* [Mac Plus keyboards](https://github.com/mamedev/mame/blob/mame0269/src/devices/bus/mackbd/pluskbd.cpp) (US, France, Japan – only key caps differ)
* [Original Mac keyboards](https://github.com/mamedev/mame/blob/mame0260/src/devices/bus/mackbd/keyboard.cpp) (US, UK, France, Italy, Japan – ANSI and ISO use different scan codes as well as different physical layout)
* [Sun Type 3/4/5 keyboards](https://github.com/mamedev/mame/blob/mame0269/src/devices/bus/sunkbd/hlekbd.cpp) (Type 5 had ANSI, ISO and JIS physical layouts – US, UK, Sweden and Japan variants confirmed)
* [Amiga keyboards](https://github.com/mamedev/mame/tree/mame0269/src/devices/bus/amiga/keyboard) (numerous language variants)

Reply:
Thanks.  I have modified the genuine Apple //e keyboard definitions to all base upon a single layout and then use PORT_MODIFY to make the minimally required alterations.  With this, future expansion for additional localized layouts will not complicate things too much.

Note though that unlike the Amiga where a US keyboard can be unplugged and French keyboard plugged in, with the Apple //e, there is just a physical switch to toggle between US and localize keyboards 'live'.  Whether we pick the US or localized layout, for the PORT_CHAR, the other one will technically be missing.   Would the emulated Apple //e keyboard have to become a 'slot' device to correctly implement both language layouts at 'Input Assignments' menu?  If it is better just to just support 1 at the UI (and natural keyboard level), then the localized one is better.

I will look at the KEYCODE_BACKSLASH2 again for my next commit.

Reply:
It’s possible to use `PORT_CONDITION` to flip the key inputs when you change the state of the switch.  Practicality depends on the number of keys that need to have their labels switched when the layout switches.

The Swiss Amiga keyboard needs to do this because a few keys switch function between French and German modes (shift states reversed): https://github.com/mamedev/mame/blob/mame0269/src/devices/bus/amiga/keyboard/matrix.cpp#L663

Reply:
Thanks.  PORT_CONDITION works if I put the condition after each PORT_BIT and create a revert PORT_BIT with the opposite condition (e.g. ISO US keyboard entries for each localized keyboard). I agree that it is only practical if there are not too many keys, but I think there may be quite a bit in the long term if more layouts are added.  I tried applying PORT_CONDITION to a 'PORT_MODIFY' line (in case it could be used to disable the entire port-modification when the condition was false, but that did not work).  For this PR, I am keeping the existing design where only the localized keyboard works in natural keyboard mode.

Regarding the Apple's ISO keyboard layouts, I tried the changes with a physical ISO PC keyboard just now, and I agree, the KEYCODE_BACKSLASH2 is needed for the <>\| key and the KEYCODE_BACKSLASH is needed for the `~£ key as you suggested.  My most recent commit makes that change.  The current version of MAME (i.e. 270) has KEYCODE_TILDE for the <>\| key.  I have not removed KEYCODE_TILDE as a mapping to that key since KEYCODE_TILDE would otherwise not be mapped to anything and I therefore expect that it be the key that those without a physical KEYCODE_BACKSLASH2 key would manually map the <>\| to.  If the correct keys are working for someone using a physical ISO keyboard, is it OK if an extra works by default too?  If not, I will remove the KEYCODE_TILDE mapping.